server/src/main/java/org/opensearch/wlm/cancellation/TaskCanceller.java

```diff
@@ -0,0 +1,244 @@
+    private final Collection<QueryGroup> activeQueryGroups;
+    private final Collection<QueryGroup> deletedQueryGroups;
+    private BooleanSupplier isNodeInDuress;
+
+    public TaskCanceller(
+        WorkloadManagementSettings workloadManagementSettings,
+        MaximumResourceTaskSelectionStrategy taskSelectionStrategy,
+        Map<String, QueryGroupLevelResourceUsageView> queryGroupLevelResourceUsageViews,
+        Collection<QueryGroup> activeQueryGroups,
+        Collection<QueryGroup> deletedQueryGroups,
+        BooleanSupplier isNodeInDuress
+    ) {
+        this.workloadManagementSettings = workloadManagementSettings;
+        this.taskSelectionStrategy = taskSelectionStrategy;
+        this.queryGroupLevelResourceUsageViews = queryGroupLevelResourceUsageViews;
+        this.activeQueryGroups = activeQueryGroups;
+        this.deletedQueryGroups = deletedQueryGroups;
+        this.isNodeInDuress = isNodeInDuress;
+    }
+
+    /**
+     * Cancel tasks based on the implemented strategy.
+     */
+    public final void cancelTasks() {
+        // cancel tasks from QueryGroups that are in Enforced mode that are breaching their resource limits
+        cancelTasks(ResiliencyMode.ENFORCED);
+        // if the node is in duress, cancel tasks accordingly.
+        handleNodeDuress();
+    }
+
+    private void handleNodeDuress() {
+        if (!isNodeInDuress.getAsBoolean()) {
+            return;
+        }
+        // List of tasks to be executed in order if the node is in duress
+        List<Consumer<Void>> duressActions = List.of(v -> cancelTasksFromDeletedQueryGroups(), v -> cancelTasks(ResiliencyMode.SOFT));
+
+        for (Consumer<Void> duressAction : duressActions) {
+            if (!isNodeInDuress.getAsBoolean()) {
+                break;
+            }
+            duressAction.accept(null);
+        }
+    }
+
+    private void cancelTasksFromDeletedQueryGroups() {
+        cancelTasks(getAllCancellableTasks(this.deletedQueryGroups));
+    }
+
+    /**
+     * Get all cancellable tasks from the QueryGroups.
+     *
+     * @return List of tasks that can be cancelled
+     */
+    List<TaskCancellation> getAllCancellableTasks(ResiliencyMode resiliencyMode) {
+        return getAllCancellableTasks(getQueryGroupsToCancelFrom(resiliencyMode));
+    }
+
+    /**
+     * Get all cancellable tasks from the given QueryGroups.
+     *
+     * @return List of tasks that can be cancelled
+     */
+    List<TaskCancellation> getAllCancellableTasks(Collection<QueryGroup> queryGroups) {
+        return queryGroups.stream().flatMap(queryGroup -> getCancellableTasksFrom(queryGroup).stream()).collect(Collectors.toList());
+    }
+
+    /**
+     * returns the list of QueryGroups breaching their resource limits.
+     *
+     * @return List of QueryGroups
+     */
+    private List<QueryGroup> getQueryGroupsToCancelFrom(ResiliencyMode resiliencyMode) {
```

Comment:
Do we not envision any use for taking multiple `resiliencyMode` as input?

Reply:
Due to the nature of how cancellation should work I think it is more reusable for different cases